src/Illuminate/Auth/SessionGuard.php

```diff
@@ -210,6 +210,71 @@ protected function userFromRecaller($recaller)
         return $user;
     }
 
+    /**
+     * Impersonate the given user.
+     *
+     * @param  \Illuminate\Contracts\Auth\Authenticatable  $user
+     * @return void
+     *
+     * @throws \Illuminate\Auth\AuthenticationException
+     */
+    public function impersonate(AuthenticatableContract $user)
+    {
+        if ($this->impersonating()) {
+            throw new AuthenticationException('Cannot impersonate while already impersonating.');
+        }
+
+        if (! $authenticated = $this->user()) {
+            throw new AuthenticationException('Cannot impersonate without a currently authenticated user.');
+        }
+
+        $this->session->put($this->getImpersonationName(), $authenticated->getAuthIdentifier());
+
+        $this->login($user);
+    }
+
+    /**
+     * Stop impersonating a user and resume the original authentication state.
+     *
+     * @return void
+     */
+    public function unpersonate()
+    {
+        if (! $id = $this->session->pull($this->getImpersonationName())) {
```

Comment:
What about stateless applications which use API only?

Reply:
This change is being proposed specifically to the `SessionGuard`, which is stateful

Reply:
Will be better to have stateless implementation also

Reply:
@siarheipashkevich I don't think it's possible (or necessary) for stateless. The impersonation would only last for the duration of the request